bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+			/* TLV must be present only once */
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+		SET_FLAG(prefix4_descr->flags,
+			 BGP_NLRI_TLV_PREFIX_DESCR_OSPF_ROUTE_TYPE);
+		prefix4_descr->ospf_route_type = pnt_decode8(&pnt);
+		break;
+	case BGP_LS_TLV_IP_REACHABILITY_INFORMATION:
+		if (CHECK_FLAG(prefix4_descr->flags,
+			       BGP_NLRI_TLV_PREFIX_DESCR_IP_REACHABILITY))
+			/* TLV must be present only once */
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+		SET_FLAG(prefix4_descr->flags,
+			 BGP_NLRI_TLV_PREFIX_DESCR_IP_REACHABILITY);
+		prefix4_descr->ip_reachability_prefixlen = pnt_decode8(&pnt);
+		memcpy(&prefix4_descr->ip_reachability_prefix.s_addr, pnt,
+		       length - sizeof(prefix4_descr
+					       ->ip_reachability_prefixlen));
+		break;
+	default:
+		assert(!"NLRI prefix4 descriptor decode cannot be used for this TLV type!");
+		break;
+	}
+
+	return BGP_NLRI_PARSE_OK;
+}
+
+static size_t bgp_linkstate_nlri_prefix4_descriptor_display(
+	struct bgp_ls_nlri_prefix4_descr_tlv *prefix4_descr, char *buf,
+	size_t size, bool multiline)
+{
+	bool first = true;
+
+	if (prefix4_descr->flags == 0)
+		return size;
+
+	snprintf(buf, size, "%sPrefix%s", multiline ? "\n" : " ",
+		 multiline ? " " : "{");
+	size -= strlen(buf);
+	buf += strlen(buf);
+
+	if (CHECK_FLAG(prefix4_descr->flags,
+		       BGP_NLRI_TLV_PREFIX_DESCR_OSPF_ROUTE_TYPE)) {
+		snprintf(buf, size, "%sOSPF-Route-Type:%u", first ? "" : " ",
+			 prefix4_descr->ospf_route_type);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(prefix4_descr->flags,
+		       BGP_NLRI_TLV_PREFIX_DESCR_IP_REACHABILITY)) {
+		snprintfrr(buf, size, "%sIPv4:%pI4/%u", first ? "" : " ",
+			   &prefix4_descr->ip_reachability_prefix,
+			   prefix4_descr->ip_reachability_prefixlen);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(prefix4_descr->flags, BGP_NLRI_TLV_PREFIX_DESCR_MT_ID)) {
+		size = bgp_linkstate_nlri_mtid_display(&prefix4_descr->mtid,
+						       buf, size, first);
+		buf += strlen(buf);
+	}
+
+	if (!multiline) {
+		snprintf(buf, size, "}");
+		size -= strlen(buf);
+	}
+
+	return size;
+}
+
+static uint16_t bgp_linkstate_nlri_prefix4_prefixlen(
+	struct bgp_linkstate_type_prefix4 *p_prefix4)
+{
+	return sizeof(struct bgp_linkstate_type_prefix4) +
+	       p_prefix4->prefix_descr.mtid.length;
+}
+
+static int
+bgp_linkstate_nlri_prefix4_decode(struct bgp_linkstate_type_prefix4 **pp,
+				  uint8_t *pnt, uint16_t len)
+{
+	struct bgp_linkstate_type_prefix4 *p_prefix4 = *pp;
+	int ret = BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+	uint8_t *lim = pnt + len;
+	uint16_t type, length;
+	uint16_t *dst, *src;
+
+	p_prefix4->proto = pnt_decode8(&pnt);
+	p_prefix4->identifier = pnt_decode64(&pnt);
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+
+		if (!bgp_ls_tlv_check_size(type, length))
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+
```

Comment:
dont return error if unknown tlv.
rfc7752: "[3.1](https://www.rfc-editor.org/rfc/rfc7752#section-3.1). Unrecognized types MUST be preserved and propagated."